cli/cppcheckexecutor.cpp

```diff
@@ -71,6 +72,117 @@
 #endif
 
 namespace {
+    class SarifReport {
+    public:
+        void addFinding(ErrorMessage msg) {
+            mFindings.push_back(std::move(msg));
+        }
+
+        picojson::array serializeRules() const {
+            picojson::array ret;
+            std::set<std::string> ruleIds;
+            for (const auto& finding : mFindings) {
+                if (ruleIds.insert(finding.id).second) {
+                    picojson::object rule;
+                    rule["id"] = picojson::value(finding.id);
+                    picojson::object shortDescription;
+                    shortDescription["text"] = picojson::value(finding.shortMessage());
+                    rule["shortDescription"] = picojson::value(shortDescription);
+                    ret.emplace_back(rule);
+                }
+            }
+            return ret;
+        }
+
+        static picojson::array serializeLocations(const ErrorMessage& finding) {
+            picojson::array ret;
+            for (const auto& location : finding.callStack) {
+                picojson::object physicalLocation;
+                picojson::object artifactLocation;
+                artifactLocation["uri"] = picojson::value(location.getOrigFile(false));
+                physicalLocation["artifactLocation"] = picojson::value(artifactLocation);
+                picojson::object region;
+                region["startLine"] = picojson::value(static_cast<int64_t>(location.line));
+                region["startColumn"] = picojson::value(static_cast<int64_t>(location.column));
+                physicalLocation["region"] = picojson::value(region);
+                picojson::object loc;
+                loc["physicalLocation"] = picojson::value(physicalLocation);
+                ret.emplace_back(loc);
+            }
+            return ret;
+        }
+
+        picojson::array serializeResults() const {
+            picojson::array results;
+            for (const auto& finding : mFindings) {
+                picojson::object res;
+                res["level"] = picojson::value(sarifLevel(finding.severity));
+                if (!finding.callStack.empty())
+                    res["locations"] = picojson::value(serializeLocations(finding));
+                picojson::object message;
+                message["text"] = picojson::value(finding.shortMessage());
+                res["message"] = picojson::value(message);
+                res["ruleId"] = picojson::value(finding.id);
+                results.emplace_back(res);
+            }
+            return results;
+        }
```

Comment:
This function does not implement the required\* `partialFingerprints` property, which is used to prevent creating duplicate alerts for the same vulnerability/finding. See [`result` object specification](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#result-object).

\* These properties are required by GitHub's Code scanning platform, not necessarily the SARIF specification.

Reply:
Also, the `locations` property is required (by GitHub), so maybe consider hoisting up the `if (!finding.callStack.empty())` condition to the for loop, so that if there are no locations, the alert is not serialized.

Reply:
ok